### PR TITLE
Make `evaluate` for universal polynomials more universal

### PR DESCRIPTION
**Pull request by SoongNoonien**
As mentioned in #2211 I've reworked the evaluation for universal polynomials. Perhaps we should add tests for the new edge cases. Evaluating at one vector always returns an element of the coercion of the provided elements in the vector an the coefficients of the polynomial. Evaluating at two vectors (one for the variable indices and one for the values) will always return a universal polynomial. All variables which aren't present yet will be ignored.

### CODE REVIEW

File: src/generic/UnivPoly.jl [hunk 1]
```diff
@@ -787,110 +787,44 @@ end
...
-   num = nvars(parent(data(a)))
-   if n > num
-      n > nvars(parent(a)) && error("Too many values")
-      if nvars(parent(data(a))) == 0
-         return constant_coefficient(data(a))*one(parent(A[1]))
-      end
-      return evaluate(data(a), A[1:num])
+function evaluate(a::UnivPoly, A::Vector{<:Union{NCRingElem, RingElement}})
+   a2 = data(a)
+   varidx = Int[var_index(x) for x in vars(a2)]
```
Comment (SoongNoonien) [inline src/generic/UnivPoly.jl]: This is not so nice I think. Ideally we would have a function called `var_indices` (or similar) which returns just the indices of the variables occurring in the given multivariate polynomial. If you think so as well, I can implement this and adjust `vars` accordingly to use `var_indices`.

File: src/generic/UnivPoly.jl [hunk 2]
```diff
@@ -904,7 +838,10 @@ function (a::Union{MPolyRingElem, UniversalPolyRingElem})(;kwargs...)
    vals = Array{RingElement}(undef, length(kwargs))
    for (i, (var, val)) in enumerate(kwargs)
      vari = findfirst(isequal(var), ss)
-     vari === nothing && error("Given polynomial has no variable $var")
+     if vari === nothing
+        isa(a, MPolyRingElem) && error("Given polynomial has no variable $var")
+        continue
+     end
```
Comment (SoongNoonien) [inline src/generic/UnivPoly.jl]: Is there a better way to achieve this? I can only think of duplicating this function for mpolys and universal polynomials.

**Comment by SoongNoonien** [comment]
Some previously failing tests were caused by the evaluation of mpoly. Apparently it is not possible to evaluate an mpoly in a ring with no variables at an empty vector with element type univariate polynomial, even though this is possible if the element type is `Int` .

**Comment by SoongNoonien** [comment]
The doctest failures with julia 1.10 seem to be unrelated.

**Comment by lgoettgens** [comment]
> Some previously failing tests were caused by the evaluation of mpoly. Apparently it is not possible to evaluate an mpoly in a ring with no variables at an empty vector with element type univariate polynomial, even though this is possible if the element type is `Int` .

Reasoning behind this: If you evaluate at an empty vector with element type univariate polynomial, the result is a univariate polynomial. But how should the result know its parent Univariate poly ring?
In the case of Int this works, as `Int` does not have a parent concept.

File: src/generic/UnivPoly.jl (hunk 1 above)
Comment (lgoettgens) [reply src/generic/UnivPoly.jl]: I generally agree, but please do that in a follow-up. Otherwise your stack of PRs waiting for one another just grows even further :)

File: test/generic/UnivPoly-test.jl [hunk 3]
```diff
@@ -978,13 +978,6 @@ end
          @test evaluate(g, V) == g([ZZ(v) for v in V]...)
          @test evaluate(g, V) == g([U(v) for v in V]...)
 
-         @test evaluate(h, V) == evaluate(h, [R(v) for v in V])
-         @test evaluate(h, V) == evaluate(h, [ZZ(v) for v in V])
-         @test evaluate(h, V) == evaluate(h, [U(v) for v in V])
-         @test evaluate(h, V) == h(V...)
-         @test evaluate(h, V) == h([ZZ(v) for v in V]...)
-         @test evaluate(h, V) == h([U(v) for v in V]...)
```
Comment (lgoettgens) [inline test/generic/UnivPoly-test.jl]: why only remove the ones for `h`? How do they differ from the ones for `g`?

File: src/generic/UnivPoly.jl (hunk 1 above)
Comment (fingolfin) [reply src/generic/UnivPoly.jl]: `var_indices` sounds like a good idea to me -- what does @thofma think?

File: src/generic/UnivPoly.jl (hunk 2 above)
Comment (fingolfin) [reply src/generic/UnivPoly.jl]: I think the code is fine as it is.

File: src/generic/UnivPoly.jl (hunk 1 above)
Comment (thofma) [reply src/generic/UnivPoly.jl]: Sounds good to me.

Comment (SoongNoonien) [reply src/generic/UnivPoly.jl]: See #2217.

File: src/generic/UnivPoly.jl [hunk 4]
```diff
@@ -900,7 +832,10 @@ function (a::Union{MPolyRingElem, UniversalPolyRingElem})(;kwargs...)
    vals = Array{RingElement}(undef, length(kwargs))
```
Comment (SoongNoonien) [inline src/generic/UnivPoly.jl]: This is not correct anymore as the number of kwargs might be greater. I'll look into it.

**Comment by SoongNoonien** [comment]
I think in a followup PR we should add documentation about the kwargs evaluation and the difference between the method for multivariate polynomials and the method of universal polynomials. The difference being that the latter ignores variable names not present in the respective parent.

**Comment by SoongNoonien** [comment]
Again. It seems to be caused by #2211. I'll rebase.
